---
title: "Cell Annotation using the Miko Scoring Pipeline"
output:
  html_document:
    theme: united
    df_print: kable
  pdf_document: default
date: 'Compiled: `r Sys.Date()`'
---


```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now, units = "secs")
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  message = FALSE,
  warning = FALSE,
  time_it = TRUE
)
```

Here we demonstrate the Miko scoring pipeline, a cell cluster scoring method that accounts for variations in gene set sizes, while simultaneously offering a hypothesis-testing framework capable of rejecting non-significantly enriched gene sets.  For a given single cell dataset, query and size-matched random gene sets are scored, and the difference between query and random module scores is scaled  using the size-adjusted standard deviation estimate obtained from a gene set-size dependent null model to yield the Miko score. Scaling by the size-adjusted standard deviation estimate corrects for size-dependencies and results in a test statistic from which a p-value can be derived. 

Using the Miko scoring pipeline, we will annotate the Human Gastrulation dataset reported by [Tyser 2021](https://www.nature.com/articles/s41586-021-04158-y). We start by reading in the data and visualizing the author-annotated population. 

```{r load data, fig.width=7, fig.height=5}

# load package
library(scMiko)

# load human gastrulation data
so.query <- readRDS("../data/demo/so_tyser2021_220621.rds")

# visualize annotated populations. 
cluster.UMAP(so = so.query, group.by = "sub_cluster") + 
  theme_void() + 
  labs(title = "Tyser 2021", subtitle = "Human Gastrulation")

```

Prior to running the Miko scoring pipeline, you must ensure that the loaded Seurat object has been clustered using `FindClusters()` and that cluster information is available in the 'seurat_clusters' meta data field. 

```{r seurat clusters}

# ensure that seurat_clusters are available
stopifnot("seurat_clusters" %in% colnames(so.query@meta.data))

# visualize seurat clusters
cluster.UMAP(so.query, group.by = "seurat_clusters")
```


## Prepare cell-type markers

Prior to performing marker-based cell-type annotation, we first prepare a comprehensive list of cell-type markers. We will use a cell-type catalog derived from multiple public scRNAseq atlases (see Cell-type marker catalog vignette for details). 

In the absence of a-priori sample information, we would include all the cell-type markers in the catalog. However, given the known early embryonic age of the profiled sample, we will proceed with cell-markers derived from murine organogenesis (Cao 2019), human fetus (Cao 2020), developing murine brain (La Manno 2021), and murine gastrulation (Pijuan-Sala 2019):


```{r cell markers}

# cell-type catalog 
marker.df <- geneSets[["Cell_Catalog"]]
marker.list <- wideDF2namedList(marker.df)

# ensure that all genes are UPPER CASE (human)
marker.list <- lapply(marker.list, toupper)

# only include gene sets with more than 3 markers
marker.list <- marker.list[unlist(lapply(marker.list, length)) > 3]

# subset markers
marker.list <- marker.list[grepl("Cao2019|Cao2020|Manno|Pijuan", names(marker.list))]

```


## Miko scoring pipeline

The miko scoring pipeline is a 3-step workflow that is implemented in `nullScore()`, `mikoScore()`, and `sigScore()` functions. 

In **step 1** we run `nullScore()`, which fits a null model that corrects for gene set-size biases. 


```{r calculate null model, fig.width=14, fig.height=4.5}

# step 1
ns.res <- nullScore(object = so.query, assay = DefaultAssay(so.query), n.replicate = 10, nbin = 24, 
                    min.gs.size = 2, max.gs.size = 200, step.size = 10, 
                    nworkers = 16, verbose = T, subsample.n = 5000)


```

From this example, the variance-gene set-size relationship can be appreciated in panel **A** and **B**. Panel **C** illustrates the correction that is applied to ensure that the null score variance is constant across all gene set sizes.  

```{r visualize null model,  fig.width=14, fig.height=4.5}

variance.mean.plot <- ns.res$variance.mean.plot
plt.null.model <- ns.res$mean.plot
plt.corrected.plot <- ns.res$corrected.plot

cowplot::plot_grid(variance.mean.plot + labs( subtitle = ""), 
                   plt.null.model, 
                   plt.corrected.plot + labs(title = "Gene Set-Size Corrected Null Scores", subtitle = ""), 
                   nrow = 1, align = "h", labels = "AUTO")
```

In **step 2**, cell-type marker sets are scored using `mikoScore()`. 

```{r cell-level scores,  fig.width=16, fig.height=15}

# step 2
so.query_scored <- mikoScore(object = so.query, geneset = marker.list, nbin = 24, 
                            nullscore = ns.res, assay = DefaultAssay(so.query), nworkers = 18)

```

Finally, in **step 3**, we identify which cell-type marker sets are significantly enriched in each cluster using `sigScore()`. 

```{r significantly enriched gene sets}

# step 3
score.result <- sigScore(object = so.query_scored,  geneset = marker.list, reduction = "umap")

# visualize 
# score.result$score_plot 

```
```{r visualize v2, echo = FALSE}

score.result$score_plot$data$raw_score <- signif(score.result$score_plot$data$raw_score, 3)
score.result$score_plot$data$miko_score <- signif(score.result$score_plot$data$miko_score, 3)
score.result$score_plot$data$significant <- score.result$score_plot$data$fdr < 0.05 & score.result$score_plot$data$miko_score > 0
score.result$score_plot$data$annotation <- paste0(score.result$score_plot$data$cluster, "_", score.result$score_plot$data$name)

plotly::ggplotly(
score.result$score_plot$data %>%
    ggplot(aes(x = raw_score, y = miko_score, 
               fill = significant,
               label = annotation)) +
    scale_fill_manual(values = c("TRUE" = "tomato", "FALSE" = "grey")) +
    geom_point(pch = 21, color = "white", size = 2) +
    theme_miko(legend = T, grid=  T) +
    geom_hline(yintercept = 0, linetype = "dashed") +
    geom_vline(xintercept = 0, linetype = "dashed") +
    labs(title = "Raw vs. Miko Enrichment Scores", x = "Raw Score", y = "Miko Score",
         subtitle = paste0(signif(sum(score.result$score_plot$data$fdr < 0.05 & 
                                        score.result$score_plot$data$miko_score > 0)/nrow(score.result$score_plot$data), 3)*100, "% Significance Rate (FDR < 0.05)"),
         fill = "FDR<0.05")
)

```

## Post-scoring filters

We use two additional post-scoring filters to fine tune which gene sets are enriched. The first is a **coherence filter** in which a positive correlation between component gene expression and the Miko score is enforced for a minimum fraction of component genes. The second is a **frequent flier filter**, which flags gene sets that exceed a minimum significance rate and represent gene sets that enrich across most cell clusters. 

### Coherence filter

```{r coh filter}

raw.mat <- so.query_scored@misc[["raw_score"]]
colnames(raw.mat) <- gsub("raw_", "", colnames(raw.mat))

df.cscore <- coherentFraction(object = so.query_scored, score.matrix =raw.mat, nworkers = 16,
                              method = "pearson",
                            genelist = marker.list, 
                            assay = DefaultAssay(so.query_scored), 
                            slot = "data", subsample.cluster.n = 500)

df.score <- score.result$cluster_stats
df.score$gs <- df.score$name
df.score$cluster <- df.score$cluster
df.merge <- merge(df.cscore, df.score, by = c("gs", "cluster"))

df.merge$sig <- df.merge$coherence_fraction >= 0.8 & df.merge$miko_score > 0 & df.merge$fdr < 0.05
plt.score.vs.coh <- df.merge %>%
  ggplot(aes(x = (miko_score), y = coherence_fraction, label = gs, fill = sig)) + 
  geom_point(pch = 21, color = "white", size = 2) + 
  theme_miko(legend = T) + 
  geom_hline(yintercept = 0) + 
  geom_vline(xintercept = 0) + 
  scale_fill_manual(values = c("TRUE" = "tomato", "FALSE" = "grey")) + 
  labs(x = "Miko Score", y = "Coherence Fraction", fill = "Enriched & Coherent\n(FDR<0.05)", 
       title = "Coherent Enrichments", 
       subtitle = paste0(100*signif(sum( df.merge$sig, na.rm = T) / nrow(df.merge), 3), "% Significance Rate"))


  plt.score.vs.coh


```


```{r consolidate results}

df.score_summary <- data.frame(cluster = df.merge$cluster, 
                              cell.type = df.merge$gs, 
                              miko_score = signif(df.merge$miko_score, 3) ,
                              p =  signif(df.merge$p, 3),
                              fdr =  signif(df.merge$fdr, 3),
                              coherence_fraction =  signif(df.merge$coherence_fraction, 3)
)

```

### Frequent flier filter

```{r flag frequent fliers}


ugrp <- unique(df.score_summary$cluster)
df.score_summary.sig <- df.score_summary %>% dplyr::filter(fdr < 0.05)
df.tally.ff <- as.data.frame(table(df.score_summary.sig$cell.type))
df.tally.ff$Freq2 <- df.tally.ff$Freq/length(ugrp)

ff.thresh <- 0.75
which.ff <- as.character(df.tally.ff$Var1[df.tally.ff$Freq2 >= ff.thresh])
df.score_summary$frequent_flier = (df.score_summary$cell.type  %in% which.ff)

```


We can visualize the frequent flier distribution to verify that an appropriate frequent flier threshold was used. In the current sample, we can see that there are a subset of marker sets that enrich indiscriminately in all clusters. In this case a threshold of 0.75 was appropriate (the threshold usually varies 0.6-0.8). 

```{r}

plt.ff <- df.tally.ff %>%
  ggplot(aes(x = Freq2)) + 
  geom_histogram(binwidth = 0.1) + 
  geom_vline(xintercept = ff.thresh, linetype = "dashed", color = "tomato") + 
  labs(title = "Frequent Flier Distribution", subtitle = paste0(signif(100*sum(df.tally.ff$Freq2 > ff.thresh)/nrow(df.tally.ff), 3), "% enriched genesets are frequent fliers (significant > ", 100*ff.thresh, "% of clusters)"),
       x = "Proportion of significant clusters", y = "Number of gene sets") + 
  theme_miko()
plt.ff

```

## Interpretting Annotations {.tabset}

Annotations are interpreted using the following statistics:

- **miko_score**: cell-type marker enrichment score. 
- **p** and **fdr**: p- and adjusted p-values (fdr) for determining significance. 
- **coherence_fraction**: Fraction of genes that are correlated to aggregate gene set expression.  
- **frequent_flier**: Flag specifying whether annotation exceeds a given significance rate (i.e., is non-specific).

Example thresholds: miko_score > 0, fdr < 0.05, coherence_fraction > 0.8, frequent_flier = F

```{r score table}
flex.asDT(df.score_summary %>% dplyr::arrange(miko_score), page_length = 10,  scrollX = TRUE)
```

The `annotationCloud()` function can be used to facilitate visualization and interpretation of annotations. 

```{r, fig.width=8, fig.height=4}

   plt.cloud <- annotationCloud(object = so.query_scored, 
                                object.group = "seurat_clusters", 
                                score = df.score_summary$miko_score,
                                score.group = df.score_summary$cluster,
                                score.cell.type = df.score_summary$cell.type,
                                score.p = df.score_summary$p,
                                score.fdr = df.score_summary$fdr,
                                score.coherence.fraction = df.score_summary$coherence_fraction,
                                score.frequent.flier = df.score_summary$frequent_flier, # 
                                fdr.correction = T, 
                                p.threshold = 0.05, 
                                coherence.threshold = 0.9, 
                                show.n.terms = 15, 
                                verbose = T)

```

### Example 1

This is an example of a clean annotation. Consistently across marker sets derived from independent studies, we find that the population is enriched for *erythroid/erythrocyte/'erythroblast* markers. Accordingly, the original authors had annotated this cluster as *erythroblasts*. 

```{r fig.width=8, fig.height=4}
plt.cloud[[9]]
```

### Example 2

In this example there is no consensus annotation. However we can see that annotations consistently identify mesenchymal-like annotations from multiple independent sources (e.g., Manno 2021 - mesenchymal progenitors; Cao 2019 - osteoblasts; Cao 2020 - smooth muscle cells; Pijuan-Sala - mesenchyme). Based on our annotations, we conclude that this cluster represents a mesenchymal population present during early human gastulation. The original authors had annotated this as *Advanced Mesoderm*. 

```{r fig.width=8, fig.height=4}
plt.cloud[[4]]
```

### Example 3

In this example the top annotations are *Distal Visceral Endoderm* and *Embyonic Definitive Endoderm* suggesting an endodermal identity. However we must be cautious when interpretting annotations using marker sets from a single source (i.e., La Manno 2021). While endoderm is not present among the other annotations, we find that *Gut* (Pijuan Sala 2019) and *Hepatoblasts* (Cao 2020) are both significantly enriched, and based on their endodermal origins, we have additional support for this cluster represents the endoderm. The original authors had annotated this cluster as *yolk sac endoderm*, *definitive endoderm* and *hypoblast*, thereby confirming our *endoderm* annotation. 

```{r fig.width=8, fig.height=4}
plt.cloud[[5]]
```


<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>

