<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Identifying Differentially Expressed Genes using the Co-Dependence Index • scMiko</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Identifying Differentially Expressed Genes using the Co-Dependence Index">
<meta property="og:description" content="scMiko">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Getting Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differentially-Expressed Genes</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Cluster Resolution Optimization</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-Type Marker Catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Cell-Type Annotation</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene Module Detection</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/R643_M01_Ochoka2021_QC_NM2_21_Sep_3_39_09_PM_2021.html">QC and Preprocessing Module</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="DEG_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Identifying Differentially Expressed Genes using the Co-Dependence Index</h1>
            
            <h4 class="date">Compiled: 2022-01-19</h4>
      
      
      <div class="hidden name"><code>DEG.Rmd</code></div>

    </div>

    
    
<div id="load-seurat-object" class="section level2">
<h2 class="hasAnchor">
<a href="#load-seurat-object" class="anchor"></a>Load Seurat object</h2>
<p>For this tutorial we will analyze the Human Gastrulation dataset reported by Tyser 2021. The dataset consists of 1195 cells. We will perform differential expression analysis using two methods: co-dependency index (CDI) and wilcoxon test.</p>
<p>We start by reading in the data and visualizing the annotated population. We are interested in identifying cell-type specific markers.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scMiko</span><span class="op">)</span>

<span class="co"># load human gastrulation data</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"../data/so_tyser2021_220621.rds"</span><span class="op">)</span>


<span class="co"># visualize clusters</span>
<span class="fu"><a href="../reference/cluster.UMAP.html">cluster.UMAP</a></span><span class="op">(</span>so <span class="op">=</span> <span class="va">so.query</span>, group.by <span class="op">=</span> <span class="st">"sub_cluster"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_void</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Tyser 2021"</span>,
    subtitle <span class="op">=</span> <span class="st">"Human Gastrulation"</span><span class="op">)</span></code></pre></div>
<p><img src="DEG_files/figure-html/load%20data-1.png" width="672"></p>
</div>
<div id="perform-differential-gene-expression-analysis" class="section level2">
<h2 class="hasAnchor">
<a href="#perform-differential-gene-expression-analysis" class="anchor"></a>Perform differential gene expression analysis</h2>
<p>To identifying cell-type specific markers, we will perform CDI- and Wilcoxon-based differentially-expressed gene (DEG) analyses. Since CDI is more computationally intensive, we will focus on the subset of genes that are expressed within the dataset.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get expressed genes</span>
<span class="va">expr_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getExpressedGenes.html">getExpressedGenes</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span><span class="op">)</span>

<span class="co"># CDI</span>
<span class="va">cdi_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/findCDIMarkers.html">findCDIMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span>, features.x <span class="op">=</span> <span class="st">"sub_cluster"</span>, features.y <span class="op">=</span> <span class="va">expr_gene</span><span class="op">)</span>

<span class="co"># Wilcoxon</span>
<span class="va">wilcoxon_output</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getDEG.html">getDEG</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span>, group_by <span class="op">=</span> <span class="st">"sub_cluster"</span>, auc.thresh <span class="op">=</span> <span class="cn">NA</span>, fdr.thresh <span class="op">=</span> <span class="cn">NA</span>,
    return.all <span class="op">=</span> <span class="cn">T</span>, return.list <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>Once the analysis has be complete, we can visualize the top differentially-expressed genes using Seurat’s <code>DotPlot</code> function.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get top markers</span>
<span class="va">wilcoxon_output.top</span> <span class="op">&lt;-</span> <span class="va">wilcoxon_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span>, wt <span class="op">=</span> <span class="va">auc</span><span class="op">)</span>
<span class="va">cdi_output</span><span class="op">$</span><span class="va">group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"sub_cluster_"</span>, <span class="st">""</span>, <span class="va">cdi_output</span><span class="op">$</span><span class="va">feature.x</span><span class="op">)</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">wilcoxon_output.top</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span>
<span class="va">cdi_output</span> <span class="op">&lt;-</span> <span class="va">cdi_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span>
<span class="va">cdi_output.top</span> <span class="op">&lt;-</span> <span class="va">cdi_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">feature.x</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">1</span>, wt <span class="op">=</span> <span class="va">ncdi</span><span class="op">)</span>

<span class="co"># generate dot plots</span>
<span class="va">plt.cdi_dot</span> <span class="op">&lt;-</span> <span class="fu">DotPlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">cdi_output.top</span><span class="op">$</span><span class="va">feature.y</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"sub_cluster"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="../reference/scale_color_miko.html">scale_color_miko</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CDI Markers"</span>, y <span class="op">=</span> <span class="st">"Clusters"</span>, x <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_miko.html">theme_miko</a></span><span class="op">(</span>legend <span class="op">=</span> <span class="cn">T</span>,
    x.axis.rotation <span class="op">=</span> <span class="fl">45</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>


<span class="va">plt.wilcoxon_dot</span> <span class="op">&lt;-</span> <span class="fu">DotPlot</span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">wilcoxon_output.top</span><span class="op">$</span><span class="va">feature</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"sub_cluster"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="../reference/scale_color_miko.html">scale_color_miko</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Wilcoxon Markers"</span>, y <span class="op">=</span> <span class="st">"Clusters"</span>, x <span class="op">=</span> <span class="st">"Genes"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_miko.html">theme_miko</a></span><span class="op">(</span>legend <span class="op">=</span> <span class="cn">T</span>,
    x.axis.rotation <span class="op">=</span> <span class="fl">45</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span>

<span class="co"># visualize</span>
<span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">plt.cdi_dot</span>, <span class="va">plt.wilcoxon_dot</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="DEG_files/figure-html/visualize-1.png" width="1056"></p>
<p>Finally, we can compare the relative sensitivity and specificity of markers obtained by each method.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get top 10 markers per cluster</span>
<span class="va">cdi_output.top</span> <span class="op">&lt;-</span> <span class="va">cdi_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">feature.x</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">ncdi</span><span class="op">)</span>
<span class="va">wilcoxon_output.top</span> <span class="op">&lt;-</span> <span class="va">wilcoxon_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">group</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">auc</span><span class="op">)</span>

<span class="co"># get sensitivity and specificity indices that were calculated by getDEG()</span>
<span class="va">cdi_sensitivity</span> <span class="op">&lt;-</span> <span class="va">wilcoxon_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">feature</span>, <span class="st">"_"</span>, <span class="va">group</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">cdi_output.top</span><span class="op">$</span><span class="va">feature.y</span>, <span class="st">"_"</span>, <span class="va">cdi_output.top</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span>
<span class="va">cdi_sensitivity</span><span class="op">$</span><span class="va">method</span> <span class="op">&lt;-</span> <span class="st">"CDI"</span>

<span class="va">wilcoxon_sensitivity</span> <span class="op">&lt;-</span> <span class="va">wilcoxon_output</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">feature</span>, <span class="st">"_"</span>, <span class="va">group</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">wilcoxon_output.top</span><span class="op">$</span><span class="va">feature</span>, <span class="st">"_"</span>, <span class="va">wilcoxon_output.top</span><span class="op">$</span><span class="va">group</span><span class="op">)</span><span class="op">)</span>
<span class="va">wilcoxon_sensitivity</span><span class="op">$</span><span class="va">method</span> <span class="op">&lt;-</span> <span class="st">"Wilcoxon"</span>

<span class="va">df.eval</span> <span class="op">&lt;-</span> <span class="fu">bind_rows</span><span class="op">(</span><span class="va">cdi_sensitivity</span>, <span class="va">wilcoxon_sensitivity</span><span class="op">)</span>

<span class="co"># generate plots</span>
<span class="va">plt_sensitivity</span> <span class="op">&lt;-</span> <span class="va">df.eval</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">method</span>, y <span class="op">=</span> <span class="va">sensitivity</span>, fill <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggbeeswarm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="../reference/theme_miko.html">theme_miko</a></span><span class="op">(</span>fill.palette <span class="op">=</span> <span class="st">"ptol"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Sensitivity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="va">plt_specificity</span> <span class="op">&lt;-</span> <span class="va">df.eval</span> <span class="op">%&gt;%</span>
    <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">method</span>, y <span class="op">=</span> <span class="va">specificity</span>, fill <span class="op">=</span> <span class="va">method</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_boxplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggbeeswarm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="../reference/theme_miko.html">theme_miko</a></span><span class="op">(</span>fill.palette <span class="op">=</span> <span class="st">"ptol"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">labs</span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Specificity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/graphics/plot.window.html">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span>

<span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html">plot_grid</a></span><span class="op">(</span><span class="va">plt_sensitivity</span>, <span class="va">plt_specificity</span><span class="op">)</span></code></pre></div>
<p><img src="DEG_files/figure-html/compare%20sensitivity%20and%20specificity-1.png" width="768"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
