<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting Started with scMiko â€¢ scMiko</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting Started with scMiko">
<meta property="og:description" content="scMiko">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="../articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="../articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting Started with scMiko</h1>
            
            <h4 data-toc-skip class="date">Compiled: 2022-08-02</h4>
      
      
      <div class="hidden name"><code>install.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p><a href="https://www.r-project.org/" class="external-link">R</a> version 4.0 or greater is
required. We also recommend installing <a href="https://www.rstudio.com/" class="external-link">R Studio</a>.</p>
<p>To install scMiko, run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Enter commands in R (or R studio)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">"NMikolajewicz/scMiko"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load scMiko</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scMiko</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="scmiko-within-the-seurat-framework">scMiko within the Seurat framework<a class="anchor" aria-label="anchor" href="#scmiko-within-the-seurat-framework"></a>
</h2>
<p>The scMiko package was developed using the <a href="https://satijalab.org/seurat/index.html" class="external-link">Seurat</a> framework.
Consequently, Seurat objects preprocessed using <a href="https://github.com/satijalab/sctransform" class="external-link">sctransform</a> are
expected in scMiko-implemented workflows.</p>
<div class="section level3">
<h3 id="option-1-preprocess-data-using-seurat-workflow">Option 1: Preprocess data using Seurat workflow<a class="anchor" aria-label="anchor" href="#option-1-preprocess-data-using-seurat-workflow"></a>
</h3>
<p>To preprocess data, users can either refer to the excellent <a href="https://satijalab.org/seurat/articles/sctransform_vignette.html" class="external-link">vignette</a>
offered by Seurat, or follow the adapted workflow shown here.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load Seurat</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>

<span class="co"># load data</span>
<span class="va">count_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/demo/ps_count_matrix.rds"</span><span class="op">)</span>
<span class="va">meta_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"../data/demo/ps_meta_data.rds"</span><span class="op">)</span>

<span class="co"># create and preprocess seurat object</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_matrix</span>, meta.data <span class="op">=</span> <span class="va">meta_data</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">so.query</span>, pattern <span class="op">=</span> <span class="st">"^mt-"</span>, col.name <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SCTransform.html" class="external-link">SCTransform</a></span><span class="op">(</span><span class="va">so.query</span>, method <span class="op">=</span> <span class="st">"glmGamPoi"</span>, vars.to.regress <span class="op">=</span> <span class="st">"percent.mt"</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">so.query</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">so.query</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">so.query</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">so.query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">so.query</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># visualize</span>
<span class="fu"><a href="../reference/cluster.UMAP.html">cluster.UMAP</a></span><span class="op">(</span><span class="va">so.query</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_void</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></code></pre></div>
<p><img src="install_files/figure-html/seurat%20object%20preprocess-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="option-2-preprocess-data-using-scpipeline-qc-preprocessing-module-">Option 2: Preprocess data using scPipeline QC &amp; Preprocessing
Module.<a class="anchor" aria-label="anchor" href="#option-2-preprocess-data-using-scpipeline-qc-preprocessing-module-"></a>
</h3>
<p>Another way to preprocess single-cell data is to use our QC &amp;
Preprocessing Module implemented in <a href="scPipeline_module_overview.html">scPipeline</a>. Given a gene x
cell count matrix, or cell ranger output, the QC &amp; Preprocessing
module applies quality-control filters and preprocesses the data. The
output from the module is a Seurat object, compatible with all
downstream analysis work flows, and an intuitive dashboard report
summarizing the data set and results.</p>
</div>
</div>
<div class="section level2">
<h2 id="sample-stratification">Sample stratification<a class="anchor" aria-label="anchor" href="#sample-stratification"></a>
</h2>
<p>Sample-level meta data is often of interest in single cell analyses.
To accommodate this information, we have included sample-wise
stratification and visualization throughout our <code>scPipeline</code>
workflows which can be leveraged by including sample-level information
in the <code>Barcodes</code> field of the Seurat object meta data. This
is an optional feature and will not affect the performance of your
analysis if omitted.</p>
</div>
<div class="section level2">
<h2 id="species-support">Species support<a class="anchor" aria-label="anchor" href="#species-support"></a>
</h2>
<p>The <code>scMiko</code> package has been optimized for <em>human</em>
and <em>murine</em> samples.</p>
<details><summary><strong>Session Info</strong>
</summary><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.2.0 (2022-04-22 ucrt)</span>
<span class="co">## Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">## Running under: Windows 10 x64 (build 19044)</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] LC_COLLATE=English_Canada.utf8  LC_CTYPE=English_Canada.utf8   </span>
<span class="co">## [3] LC_MONETARY=English_Canada.utf8 LC_NUMERIC=C                   </span>
<span class="co">## [5] LC_TIME=English_Canada.utf8    </span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">## [1] scMiko_0.1.0        flexdashboard_0.5.2 tidyr_1.2.0        </span>
<span class="co">## [4] sp_1.4-7            SeuratObject_4.1.0  Seurat_4.1.1       </span>
<span class="co">## [7] dplyr_1.0.9         ggplot2_3.3.6      </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] systemfonts_1.0.4           plyr_1.8.7                 </span>
<span class="co">##   [3] igraph_1.3.1                lazyeval_0.2.2             </span>
<span class="co">##   [5] splines_4.2.0               listenv_0.8.0              </span>
<span class="co">##   [7] scattermore_0.8             GenomeInfoDb_1.32.2        </span>
<span class="co">##   [9] digest_0.6.29               htmltools_0.5.2            </span>
<span class="co">##  [11] fansi_1.0.3                 magrittr_2.0.3             </span>
<span class="co">##  [13] memoise_2.0.1               tensor_1.5                 </span>
<span class="co">##  [15] cluster_2.1.3               ROCR_1.0-11                </span>
<span class="co">##  [17] globals_0.15.0              matrixStats_0.62.0         </span>
<span class="co">##  [19] pkgdown_2.0.3               spatstat.sparse_2.1-1      </span>
<span class="co">##  [21] colorspace_2.0-3            ggrepel_0.9.1              </span>
<span class="co">##  [23] textshaping_0.3.6           xfun_0.31                  </span>
<span class="co">##  [25] RCurl_1.98-1.6              crayon_1.5.1               </span>
<span class="co">##  [27] jsonlite_1.8.0              progressr_0.10.0           </span>
<span class="co">##  [29] spatstat.data_2.2-0         survival_3.3-1             </span>
<span class="co">##  [31] zoo_1.8-10                  glue_1.6.2                 </span>
<span class="co">##  [33] polyclip_1.10-0             gtable_0.3.0               </span>
<span class="co">##  [35] zlibbioc_1.42.0             XVector_0.36.0             </span>
<span class="co">##  [37] leiden_0.4.2                DelayedArray_0.22.0        </span>
<span class="co">##  [39] future.apply_1.9.0          BiocGenerics_0.42.0        </span>
<span class="co">##  [41] abind_1.4-5                 scales_1.2.0               </span>
<span class="co">##  [43] DBI_1.1.2                   spatstat.random_2.2-0      </span>
<span class="co">##  [45] miniUI_0.1.1.1              Rcpp_1.0.8.3               </span>
<span class="co">##  [47] viridisLite_0.4.0           xtable_1.8-4               </span>
<span class="co">##  [49] reticulate_1.25             spatstat.core_2.4-4        </span>
<span class="co">##  [51] stats4_4.2.0                htmlwidgets_1.5.4          </span>
<span class="co">##  [53] httr_1.4.3                  RColorBrewer_1.1-3         </span>
<span class="co">##  [55] ellipsis_0.3.2              ica_1.0-2                  </span>
<span class="co">##  [57] farver_2.1.0                pkgconfig_2.0.3            </span>
<span class="co">##  [59] sass_0.4.1                  uwot_0.1.11                </span>
<span class="co">##  [61] deldir_1.0-6                utf8_1.2.2                 </span>
<span class="co">##  [63] labeling_0.4.2              tidyselect_1.1.2           </span>
<span class="co">##  [65] rlang_1.0.2                 reshape2_1.4.4             </span>
<span class="co">##  [67] later_1.3.0                 munsell_0.5.0              </span>
<span class="co">##  [69] tools_4.2.0                 cachem_1.0.6               </span>
<span class="co">##  [71] cli_3.3.0                   generics_0.1.2             </span>
<span class="co">##  [73] ggridges_0.5.3              evaluate_0.15              </span>
<span class="co">##  [75] stringr_1.4.0               fastmap_1.1.0              </span>
<span class="co">##  [77] yaml_2.3.5                  ragg_1.2.2                 </span>
<span class="co">##  [79] goftest_1.2-3               knitr_1.39                 </span>
<span class="co">##  [81] fs_1.5.2                    fitdistrplus_1.1-8         </span>
<span class="co">##  [83] purrr_0.3.4                 RANN_2.6.1                 </span>
<span class="co">##  [85] sparseMatrixStats_1.8.0     pbapply_1.5-0              </span>
<span class="co">##  [87] future_1.25.0               nlme_3.1-157               </span>
<span class="co">##  [89] mime_0.12                   formatR_1.12               </span>
<span class="co">##  [91] compiler_4.2.0              rstudioapi_0.13            </span>
<span class="co">##  [93] plotly_4.10.0               png_0.1-7                  </span>
<span class="co">##  [95] spatstat.utils_2.3-1        tibble_3.1.7               </span>
<span class="co">##  [97] bslib_0.3.1                 glmGamPoi_1.8.0            </span>
<span class="co">##  [99] stringi_1.7.6               highr_0.9                  </span>
<span class="co">## [101] RSpectra_0.16-1             desc_1.4.1                 </span>
<span class="co">## [103] rgeos_0.5-9                 lattice_0.20-45            </span>
<span class="co">## [105] Matrix_1.4-1                vctrs_0.4.1                </span>
<span class="co">## [107] pillar_1.7.0                lifecycle_1.0.1            </span>
<span class="co">## [109] spatstat.geom_2.4-0         lmtest_0.9-40              </span>
<span class="co">## [111] jquerylib_0.1.4             RcppAnnoy_0.0.19           </span>
<span class="co">## [113] bitops_1.0-7                data.table_1.14.2          </span>
<span class="co">## [115] cowplot_1.1.1               irlba_2.3.5                </span>
<span class="co">## [117] GenomicRanges_1.48.0        httpuv_1.6.5               </span>
<span class="co">## [119] patchwork_1.1.1             R6_2.5.1                   </span>
<span class="co">## [121] promises_1.2.0.1            KernSmooth_2.23-20         </span>
<span class="co">## [123] gridExtra_2.3               IRanges_2.30.0             </span>
<span class="co">## [125] parallelly_1.31.1           codetools_0.2-18           </span>
<span class="co">## [127] MASS_7.3-57                 assertthat_0.2.1           </span>
<span class="co">## [129] SummarizedExperiment_1.26.1 rprojroot_2.0.3            </span>
<span class="co">## [131] withr_2.5.0                 sctransform_0.3.3          </span>
<span class="co">## [133] GenomeInfoDbData_1.2.8      S4Vectors_0.34.0           </span>
<span class="co">## [135] mgcv_1.8-40                 parallel_4.2.0             </span>
<span class="co">## [137] grid_4.2.0                  rpart_4.1.16               </span>
<span class="co">## [139] DelayedMatrixStats_1.18.0   rmarkdown_2.14             </span>
<span class="co">## [141] MatrixGenerics_1.8.0        Rtsne_0.16                 </span>
<span class="co">## [143] Biobase_2.56.0              shiny_1.7.1</span></code></pre>
</details>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=" class="external-link">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
