<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scPipeline analysis workflows â€¢ scMiko</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scPipeline analysis workflows">
<meta property="og:description" content="scMiko">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="../articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="../articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scPipeline_analysis_workflows_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scPipeline analysis workflows</h1>
            
            <h4 class="date">Compiled: 2022-01-27</h4>
      
      
      <div class="hidden name"><code>scPipeline_analysis_workflows.Rmd</code></div>

    </div>

    
    
<p>The following workflows are outlined in this section:</p>
<ol style="list-style-type: decimal">
<li>Prepare data for custom downstream analysis</li>
<li>Determine cluster-Specific Genes</li>
<li>Annotate cell types</li>
<li>Evaluate gene set expression(s) and associations</li>
<li>Identify gene modules</li>
</ol>
<div id="prepare-data-for-custom-downstream-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#prepare-data-for-custom-downstream-analysis" class="anchor"></a>1. Prepare data for custom downstream analysis</h3>
<ol style="list-style-type: decimal">
<li>
<strong>M01: Pre-process data</strong>
<ul>
<li>Generates Seurat object with normalized and scaled single-cell data.</li>
</ul>
</li>
<li>
<strong>M02: Integrate data</strong> (<em>optional</em>)
<ul>
<li>Only necessary if batch correction and integration is required for multiple data sets.</li>
</ul>
</li>
<li>
<strong>M18: Identify optimal cluster resolution</strong> (<em>optional</em>)
<ul>
<li>Evaluate cell population clusters at several resolutions to identify optimal resolution.</li>
</ul>
</li>
</ol>
</div>
<div id="determine-cluster-specific-genes" class="section level3">
<h3 class="hasAnchor">
<a href="#determine-cluster-specific-genes" class="anchor"></a>2. Determine cluster-specific genes</h3>
<ol style="list-style-type: decimal">
<li>
<strong>M01: Pre-process data</strong>
<ul>
<li>Generates Seurat object with normalized and scaled single-cell data.</li>
</ul>
</li>
<li>
<strong>M02: Integrate data</strong> (<em>optional</em>)
<ul>
<li>Only necessary if batch correction and integration is required for multiple data sets.</li>
</ul>
</li>
<li>
<strong>M18: Identify optimal cluster resolution</strong>
<ul>
<li>Evaluate cell population clusters at several resolutions to identify optimal resolution.</li>
<li>Differential genes are computed for each evaluated cluster using <em>Wilcoxon</em> and <em>CDI</em> DE methods.</li>
</ul>
</li>
</ol>
</div>
<div id="annotate-cell-types" class="section level3">
<h3 class="hasAnchor">
<a href="#annotate-cell-types" class="anchor"></a>3. Annotate cell types</h3>
<ol style="list-style-type: decimal">
<li>
<strong>M01: Pre-process data</strong>
<ul>
<li>Generates Seurat object with normalized and scaled single-cell data.</li>
</ul>
</li>
<li>
<strong>M02: Integrate data</strong> (<em>optional</em>)
<ul>
<li>Only necessary if batch correction and integration is required for multiple data sets.</li>
</ul>
</li>
<li>
<strong>M18: Identify optimal cluster resolution</strong> (<em>optional</em>)
<ul>
<li>Evaluate cell population clusters at several resolutions to identify optimal resolution.</li>
</ul>
</li>
<li>
<strong>M05: Annotate cell types</strong>
<ul>
<li>Annotate clusters at resolution identified in M18.</li>
</ul>
</li>
</ol>
</div>
<div id="evaluate-gene-expressions-and-associations" class="section level3">
<h3 class="hasAnchor">
<a href="#evaluate-gene-expressions-and-associations" class="anchor"></a>4. Evaluate gene expression(s) and associations</h3>
<ol style="list-style-type: decimal">
<li>
<strong>M01: Pre-process data</strong>
<ul>
<li>Generates Seurat object with normalized and scaled single-cell data.<br>
</li>
</ul>
</li>
<li>
<strong>M02: Integrate data</strong> (<em>optional</em>)
<ul>
<li>Only necessary if batch correction and integration is required for multiple data sets.</li>
</ul>
</li>
<li>
<strong>M18: Identify optimal cluster resolution</strong> (<em>optional</em>)
<ul>
<li>Evaluate cell population clusters at several resolutions to identify optimal resolution.</li>
</ul>
</li>
<li><strong>M09: Evaluate gene expression and association profiles</strong></li>
</ol>
</div>
<div id="identify-gene-modules" class="section level3">
<h3 class="hasAnchor">
<a href="#identify-gene-modules" class="anchor"></a>5. Identify gene modules</h3>
<ol style="list-style-type: decimal">
<li>
<strong>M01: Pre-process data</strong>
<ul>
<li>Generates Seurat object with normalized and scaled single-cell data.</li>
</ul>
</li>
<li>
<strong>M02: Integrate data</strong> (<em>optional</em>)
<ul>
<li>Only necessary if batch correction and integration is required for multiple data sets.</li>
</ul>
</li>
<li>
<strong>M18: Identify optimal cluster resolution</strong> (<em>optional</em>)
<ul>
<li>Evaluate cell population clusters at several resolutions to identify optimal resolution.</li>
</ul>
</li>
<li>
<strong>M24: Identify gene modules</strong>.</li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
