<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>scPipeline analysis workflows • scMiko</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="scPipeline analysis workflows">
<meta property="og:description" content="scMiko">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="../articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="../articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="scPipeline_module_overview_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>scPipeline analysis workflows</h1>
            
            <h4 class="date">Compiled: 2022-01-25</h4>
      
      
      <div class="hidden name"><code>scPipeline_module_overview.Rmd</code></div>

    </div>

    
    
<hr>
<p>In this article we overview each scPipeline module, and provide details on how to specify parameters for analysis. For each module, we provide a description, example reports and documentation for implementation (i.e., parameter specification).</p>
<p>Prior to running any scPipeline module, please visit our <a href="scPipeline_getting_started.html">getting started with scPipeline</a> article to setup up scPipeline.</p>
<div id="scpipeline-parameter-specification" class="section level3">
<h3 class="hasAnchor">
<a href="#scpipeline-parameter-specification" class="anchor"></a>scPipeline Parameter specification</h3>
<p>In general, our convention is to store all analysis parameters in a list named <code>parameter.list</code> at the start of the script.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># example</span>
<span class="va">parameter.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>input.file <span class="op">=</span> <span class="st">"../data.rds"</span>, <span class="va">...</span>, print.inline <span class="op">=</span> <span class="cn">F</span>, save.flag <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p>In all scPipeline modules, <code>parameter.list</code> is already specified, so users need only to configure the analysis parameters (e.g., specify input data, set cluster resolution, etc.) prior to running the module. In general, the default settings will yield reasonable results, however a little time spent optimizing parameters will often yield better outcomes.</p>
<hr>
</div>
<div id="module-01-quality-control-qc-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#module-01-quality-control-qc-preprocessing" class="anchor"></a>Module 01: Quality Control (QC) &amp; Preprocessing</h2>
<div id="description" class="section level3">
<h3 class="hasAnchor">
<a href="#description" class="anchor"></a>Description</h3>
<p>Initial appraisal of scRNA-seq count matrix data and overview of quality control metrics, data filtering, and normalization. Generates initial UMAP clusters, and PCA analyses. M01 output saves seurat object.</p>
</div>
<div id="example-reports" class="section level3">
<h3 class="hasAnchor">
<a href="#example-reports" class="anchor"></a>Example Reports</h3>
<ul>
<li><a href="Reports/M01_QC_demo_cao2019_10000cells_240122.html">Cao 2019: Murine organogenesis</a></li>
<li><a href="Reports/M01_QC_demo_manno_10000cells_240122.html">La Manno 2021: Developing murine brain</a></li>
<li><a href="Reports/M01_QC_demo_ps_10000cells_240122.html">Pijuan-Sala 2019: Murine gastrulation</a></li>
<li><a href="Reports/M01_QC_demo_zeisel_10000cells_240122.html">Zeisel 2018: Adolescent murine brain</a></li>
</ul>
<p>All datasets were subsampled to 10,000 cells and minimal cell filtering was performed due to original count matrices having been cleaned by authors.</p>
</div>
<div id="inputs-and-outputs" class="section level3">
<h3 class="hasAnchor">
<a href="#inputs-and-outputs" class="anchor"></a>Inputs and Outputs</h3>
<details><p><summary><strong>show contents</strong></summary></p>
<p><u>Inputs</u></p>
<p>Any of the following data formats is expected as an input:</p>
<ul>
<li>
<strong>Seurat object</strong>: must contain RNA assay with gene x cell count matrix</li>
<li>
<strong>Cell Ranger output</strong>: folder must contain <code>barcodes.tsv</code>, <code>genes.tsv</code>, <code>matrix.mtx</code>
</li>
<li><strong>gene x cell count matrix</strong></li>
</ul>
<p><u>Outputs</u></p>
<ul>
<li>
<strong>Seurat object</strong> (.Rdata file)</li>
<li>
<strong>Analysis report</strong> (.HTML file)</li>
</ul></details>
</div>
<div id="analysis-parameters" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-parameters" class="anchor"></a>Analysis Parameters</h3>
<details><p><summary><strong>show table</strong></summary></p>
<table class="table">
<colgroup>
<col width="11%">
<col width="55%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Argument</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>input.object</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>input.cell_ranger</strong></td>
<td>path to folder containing cell ranger output</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>input.matrix</strong></td>
<td>path gene x cell count matrix (.tsv, .csv format)</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>save.flag</strong></td>
<td>save Seurat object from current analysis</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>output.object.directory</strong></td>
<td>path to folder in which to save Seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>output.object.file</strong></td>
<td>name of file to which Seurat object is saved. Must have “.Rdata” suffix (e.g., ‘object.Rdata’)</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>subsample.n</strong></td>
<td>value specifying how many cells to sub sample. Set to NA if none.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>cluster.resolution</strong></td>
<td>value of the resolution parameter passed to <code><a href="https://satijalab.org/seurat/reference/FindClusters.html">Seurat::FindClusters(...)</a></code>. Use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities. Default is 1.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>gene.upperlimit</strong></td>
<td>Upper limit of genes per cell. Value between [0, Inf], but more than <code>gene.lowerlimit</code>. Any cells exceeding this threshold are filtered out. Default is 9000.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>gene.lowerlimit</strong></td>
<td>Lower limit of genes per cell. Value between [0, Inf], but less than <code>gene.upperlimit</code>. Any cells below this threshold are filtered out. Default is 200.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>mt.upperlimit</strong></td>
<td>Upper limit of mitochondrial content per cell. Value between [0, 100]. Cells exceeding this threshold are filtered out. Default is 10.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>unmatched.rate.filter.flag</strong></td>
<td>Logical specifying whether to filtered cells using unmatched rates. Default is TRUE.</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>vars2regress</strong></td>
<td>Variables to regress out during scaling step. Default is ‘percent.mt’.</td>
<td>vector of character(s)</td>
</tr>
<tr class="even">
<td><strong>correct.artifact</strong></td>
<td>identify artifact cells and omit from dimensional reduction. Default is TRUE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>feature.select.method</strong></td>
<td>Feature selection method for dimensional reduction.</td>
<td>- hvg (default): highly-variable genes <br>- deviance</td>
</tr>
<tr class="even">
<td><strong>scale.method</strong></td>
<td>Expression scaling method.</td>
<td>- sct (default): SCTransform residuals <br>- null_residuals</td>
</tr>
<tr class="odd">
<td><strong>pca.method</strong></td>
<td>principal component analysis method.</td>
<td>- pca (default) <br>- glmpca</td>
</tr>
<tr class="even">
<td><strong>pca.component.select.method</strong></td>
<td>method for selecting number of principal components for clustering and UMAP embedding.</td>
<td>- cum_var (default) <br>- elbow</td>
</tr>
<tr class="odd">
<td><strong>pca.var.threshold</strong></td>
<td>value between [0, 1] specifying cumulative variance explained by pca. Ignored if <code>pca.component.select.method</code> is not ‘cum_var’.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>conserve.memory</strong></td>
<td>If TRUE, the residual matrix for all genes computed using SCTransform is never created in full; useful for large data sets, but will take longer to run. Default is FALSE</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>species.filter.flag</strong></td>
<td>Logical specifying whether to filter species. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>species.include</strong></td>
<td>Specify which species to retain in analysis. Ignored if <code>species.filter.flag</code> is FALSE</td>
<td>- Hs: human <br>- Mm: murine</td>
</tr>
<tr class="odd">
<td><strong>nworkers</strong></td>
<td>Number of workers used for parallel implementation of SCTransform. Default is 1.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>save.pdf</strong></td>
<td>Logical specifying whether to save figures and tables in separate directory.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>print.inline</strong></td>
<td>print figures after each code chunk. Recommended if running script chunk-by-chunk. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>developer</strong></td>
<td>Developer mode. Default is FALSE.</td>
<td>logical</td>
</tr>
</tbody>
</table></details><hr>
</div>
</div>
<div id="module-02-data-integration" class="section level2">
<h2 class="hasAnchor">
<a href="#module-02-data-integration" class="anchor"></a>Module 02: Data Integration</h2>
<div id="description-1" class="section level3">
<h3 class="hasAnchor">
<a href="#description-1" class="anchor"></a>Description</h3>
<p>Batch correction and integration of independent experiments. Generates a new Seurat <code>integrated</code> assay using one of two approaches:</p>
<ol style="list-style-type: decimal">
<li>
<strong>rPCA</strong>: reciprocal PCA integration
<ul>
<li>Recommended for larger datasets</li>
</ul>
</li>
<li>
<strong>CCA</strong>: canonical correlation analysis
<ul>
<li>Recommended method, however computationally demanding and incompatible with larger data</li>
</ul>
</li>
</ol>
<p>See <a href="https://www.cell.com/cell/fulltext/S0092-8674(19)30559-8">Stuart et al. (2019). Cell</a> for detailed description.</p>
<p>The primary aim of data integration is to identify common cell populations between datasets. However, because each method involves non-linear warping, many statistical assumptions are broken in the process, and thus expression matrix from the <code>integrated</code> assay is not appropriate for downstream differential gene expression. Specifically, the <code>integrated</code> assay should only be used for clustering and visualization. If downstream differential expression (DEG) analysis is performed, it is recommended that the existing <code>RNA</code> assay is normalized and scaled.</p>
<hr>
</div>
</div>
<div id="module-03-cluster-optimization" class="section level2">
<h2 class="hasAnchor">
<a href="#module-03-cluster-optimization" class="anchor"></a>Module 03: Cluster Optimization</h2>
<div id="description-2" class="section level3">
<h3 class="hasAnchor">
<a href="#description-2" class="anchor"></a>Description</h3>
<p>Clustering is performed at several candidate resolutions, and the optimal resolution is informed using:</p>
<ul>
<li>Marker specificity analysis (CDI-based)</li>
<li>Silhouette analysis.</li>
<li>Nearest-neighbor purity analysis</li>
<li>Differential gene expression (Wilcoxon, CDI)</li>
</ul>
</div>
<div id="example-reports-1" class="section level3">
<h3 class="hasAnchor">
<a href="#example-reports-1" class="anchor"></a>Example Reports</h3>
<ul>
<li><a href="Reports/M18_ClusterOptimization_cao2019_10000cells_240122.html">Cao 2019: Murine organogenesis</a></li>
<li><a href="Reports/M18_ClusterOptimization_manno_10000cells_240122.html">La Manno 2021: Developing murine brain</a></li>
<li><a href="Reports/M18_ClusterOptimization_ps_10000cells_240122.html">Pijuan-Sala 2019: Murine gastrulation</a></li>
<li><a href="Reports/M18_ClusterOptimization_zeisel_10000cells_240122.html">Zeisel 2018: Adolescent murine brain</a></li>
</ul>
</div>
<div id="inputs-and-outputs-1" class="section level3">
<h3 class="hasAnchor">
<a href="#inputs-and-outputs-1" class="anchor"></a>Inputs and Outputs</h3>
<details><p><summary><strong>show contents</strong></summary></p>
<p><u>Inputs</u></p>
<ul>
<li>
<strong>Seurat object</strong> (.Rds or .RData file)</li>
</ul>
<p><u>Outputs</u></p>
<ul>
<li>
<strong>Analysis report</strong> (.HTML file)</li>
</ul></details>
</div>
<div id="analysis-parameters-1" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-parameters-1" class="anchor"></a>Analysis Parameters</h3>
<details><p><summary><strong>show table</strong></summary></p>
<table class="table">
<colgroup>
<col width="11%">
<col width="55%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Argument</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>input.file</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>cluster.resolution</strong></td>
<td>range of resolution values used for evaluating candidate cluster configurations. Recommended = c(0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2, 3)</td>
<td>vector of numeric(s)</td>
</tr>
<tr class="odd">
<td><strong>lfc.threshold</strong></td>
<td>log-fold change threshold used in differential-expression analysis. 0.5 is recommended.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>fdr.threshold</strong></td>
<td>fdr-value significance threshold. Value between [0,1]. Default is 0.05.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>group.singletons</strong></td>
<td>Group singletons into nearest cluster. If FALSE, assign all singletons to a “singleton” group. For certain difficult-to-cluster data sets, <code>group.singletons</code> = FALSE may yield faster results. Default is FALSE</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>print.inline</strong></td>
<td>print figures after each code chunk. Recommended if running script chunk-by-chunk. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>subsample_factor</strong></td>
<td>Value between [0,1] specifying fraction of cells to subsample. Use 1 for no subsampling.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>subsample_n</strong></td>
<td>value specifying how many cells to sub sample. Set to NA if none.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>n.workers</strong></td>
<td>number of workers used for parallel implementation. <code><a href="https://rdrr.io/r/parallel/detectCores.html">parallel::detectCores()</a></code> is recommended.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>only.pos</strong></td>
<td>only included genes with logFC &gt; 0. Default is TRUE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>save.pdf</strong></td>
<td>save figures and tables</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>update.log</strong></td>
<td>update analysis log. Ignored if developer = F.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>rprofile.dir</strong></td>
<td>look for input file in path specified in .Rprofile. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>developer</strong></td>
<td>Developer mode. Default is FALSE.</td>
<td>logical</td>
</tr>
</tbody>
</table></details><hr>
</div>
</div>
<div id="module-04-cell-type-annotation" class="section level2">
<h2 class="hasAnchor">
<a href="#module-04-cell-type-annotation" class="anchor"></a>Module 04: Cell-Type Annotation</h2>
<div id="description-3" class="section level3">
<h3 class="hasAnchor">
<a href="#description-3" class="anchor"></a>Description</h3>
<p>Annotation of cell populations identified by unsupervised clustering. Cell-type markers are scored using the Miko scoring pipeline.</p>
<p>Prior to running this module, we recommend running the cluster optimization module to identify the appropriate resolution for clustering.</p>
</div>
<div id="example-reports-2" class="section level3">
<h3 class="hasAnchor">
<a href="#example-reports-2" class="anchor"></a>Example Reports</h3>
<ul>
<li><a href="Reports/M05_CellAnnotation_demo_cao2019_10000cells_240122.html">Cao 2019: Murine organogenesis</a></li>
<li><a href="Reports/M05_CellAnnotation_demo_manno_10000cells_240122.html">La Manno 2021: Developing murine brain</a></li>
<li><a href="Reports/M05_CellAnnotation_demo_ps_10000cells_240122.html">Pijuan-Sala 2019: Murine gastrulation</a></li>
<li><a href="Reports/M05_CellAnnotation_demo_zeisel_10000cells_240122.html">Zeisel 2018: Adolescent murine brain</a></li>
</ul>
</div>
<div id="inputs-and-outputs-2" class="section level3">
<h3 class="hasAnchor">
<a href="#inputs-and-outputs-2" class="anchor"></a>Inputs and Outputs</h3>
<details><p><summary><strong>show contents</strong></summary></p>
<p><u>Inputs</u></p>
<ul>
<li>
<strong>Seurat object</strong> (.Rds or .RData file)</li>
</ul>
<p><u>Outputs</u></p>
<ul>
<li>
<strong>Analysis report</strong> (.HTML file)</li>
</ul></details>
</div>
<div id="analysis-parameters-2" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-parameters-2" class="anchor"></a>Analysis Parameters</h3>
<details><p><summary><strong>show table</strong></summary></p>
<table class="table">
<colgroup>
<col width="11%">
<col width="55%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Argument</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>input.file</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>external.markers</strong></td>
<td>cell-type markers used for annotation</td>
<td>named list; names are cell-types, entries are genes. <br>E.g., list(cell_A = c(“gene1”, “gene2”, “gene3”), cell_B = c(“gene4”, “gene5”, “gene6”))</td>
</tr>
<tr class="odd">
<td><strong>internal.markers</strong></td>
<td>name of scMiko internal gene set(s). Run <code><a href="https://rdrr.io/r/base/names.html">names(geneSets)</a></code> for list of available gene sets.</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>marker.filter</strong></td>
<td>filter marker sets using <code>grepl</code> pattern matching. E.g., Manno2021 for all La Manno 2021 atlas-derived markers in <code>Cell_Catalogy</code> stored in scMiko <code>geneSets</code>.</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>coherence.threshold</strong></td>
<td>value of minimum gene set coherence. Gene sets with coherence scores below this threshold are disqualified from consideration. Value must be between [0,1]. Default is 0.8.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>p.threshold</strong></td>
<td>p-value significance threshold. Value between [0,1]. Default is 0.05.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>frequent.filter.threshold</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>filter.frequent.fliers</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>min.representation</strong></td>
<td>minimum fraction of genes required for gene set consideration. If fraction of genes present in seurat object is below this threshold, gene set is disqualified. Value between [0,1]. Default is 0.5.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>cluster.resolution</strong></td>
<td>value of the resolution parameter passed to <code><a href="https://satijalab.org/seurat/reference/FindClusters.html">Seurat::FindClusters(...)</a></code>. Use a value above (below) 1.0 if you want to obtain a larger (smaller) number of communities. Default is 1.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>fdr.correction</strong></td>
<td>apply Benjamini &amp; Hochberg correction</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>clean.clusters</strong></td>
<td>filter cells that exceed 3 median absolute deviations from cluster center in UMAP coordinate space. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>max.cells</strong></td>
<td>max number of cells used in Miko scoring pipeline. Default is 20000.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>subsample_factor</strong></td>
<td>value specifying fraction of cells to sub sample. Value between [0,1]</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>min.geneset.size</strong></td>
<td>maximal gene set size. Default is 200.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>max.geneset.size</strong></td>
<td>minimal gene set size. Default is 3.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>pathway.db</strong></td>
<td>database used for functional annotation of gene sets</td>
<td>- GO <br>- Bader</td>
</tr>
<tr class="even">
<td><strong>n.workers</strong></td>
<td>number of workers used for parallel implementation. Default is <code><a href="https://rdrr.io/r/parallel/detectCores.html">parallel::detectCores()</a></code>.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>save.pdf</strong></td>
<td>save figures and tables.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>show.top.n.gene.panels</strong></td>
<td>max number of gene sets per cluster to show in report. Default is 5</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>max.gene.panels.allowed</strong></td>
<td>max number of overall gene sets to show in report. Default is 75</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>print.inline</strong></td>
<td>print figures after each code chunk. Recommended if running script chunk-by-chunk. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>update.log</strong></td>
<td>update analysis log. Ignored if developer = F.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>rprofile.dir</strong></td>
<td>look for input file in path specified in .Rprofile. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>developer</strong></td>
<td>Developer mode. Default is FALSE.</td>
<td>logical</td>
</tr>
</tbody>
</table></details><hr>
</div>
</div>
<div id="module-05-gene-program-discovery" class="section level2">
<h2 class="hasAnchor">
<a href="#module-05-gene-program-discovery" class="anchor"></a>Module 05: Gene Program Discovery</h2>
<div id="description-4" class="section level3">
<h3 class="hasAnchor">
<a href="#description-4" class="anchor"></a>Description</h3>
<p>Gene program/module identification. Three independent methods are implemented to identify gene programs based on gene co-expression profiles:</p>
<ul>
<li>Scale-free shared nearest neighbor network analysis (SSN)</li>
<li>Independent component analysis (ICA)</li>
<li>Non-negative matrix factorization (NMF)</li>
</ul>
<p>Gene set enrichment is performed on each gene program to facilitate functional annotation.</p>
</div>
<div id="example-reports-3" class="section level3">
<h3 class="hasAnchor">
<a href="#example-reports-3" class="anchor"></a>Example Reports</h3>
<ul>
<li><a href="Reports/M05_GeneProgramDiscovery_demo_cao2019_10000cells_250122.html">Cao 2019: Murine organogenesis</a></li>
<li><a href="Reports/M05_GeneProgramDiscovery_demo_manno_10000cells_250122.html">La Manno 2021: Developing murine brain</a></li>
<li><a href="Reports/M05_GeneProgramDiscovery_demo_ps_10000cells_250122.html">Pijuan-Sala 2019: Murine gastrulation</a></li>
<li><a href="Reports/M05_GeneProgramDiscovery_demo_zeisel_10000cells_250122.html">Zeisel 2018: Adolescent murine brain</a></li>
</ul>
</div>
<div id="inputs-and-outputs-3" class="section level3">
<h3 class="hasAnchor">
<a href="#inputs-and-outputs-3" class="anchor"></a>Inputs and Outputs</h3>
<details><p><summary><strong>show contents</strong></summary></p>
<p><u>Inputs</u></p>
<ul>
<li>
<strong>Seurat object</strong> (.Rds or .RData file)</li>
</ul>
<p><u>Outputs</u></p>
<ul>
<li>
<strong>Analysis report</strong> (.HTML file)</li>
</ul></details>
</div>
<div id="analysis-parameters-3" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-parameters-3" class="anchor"></a>Analysis Parameters</h3>
<details><p><summary><strong>show table</strong></summary></p>
<table class="table">
<colgroup>
<col width="11%">
<col width="55%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Argument</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>input.file</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>cluster.resolution</strong></td>
<td>Value of the resolution parameter, use a value above 1.0 if you want to obtain a larger number of communities.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>subsample_factor</strong></td>
<td>Value between [0,1] specifying fraction of cells to subsample. Use 1 for no subsampling.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>subsample_n</strong></td>
<td>value specifying how many cells to sub sample. Set to NA if none.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>subset.data</strong></td>
<td>Subset Seurat object according to specified meta data field. Only specified meta data entries are retained, while remaining of data is omitted. Set as “no.subset” for not subsetting.</td>
<td>data frame specifying which field (<code>df$field</code>) to subset on, and which field entries to retain (<code>df$subgroups</code>)</td>
</tr>
<tr class="even">
<td><strong>batch.feature</strong></td>
<td>meta data field to use for batch correction (set to NA if unspecified)</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>feature.selection.method</strong></td>
<td>method used to select features for module detection analysis.</td>
<td>- expr: expression-based criterion <br>- hvg: highly-variable genes <br>- deviance: deviance-based criterion</td>
</tr>
<tr class="even">
<td><strong>max.features</strong></td>
<td>max number of features to include in module detection analysis (NA if unspecified)</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>min.pct</strong></td>
<td>minimal feature expression (applicable only if <code>select.method</code> = “expr”). Value between [0,1], use 0.1 for large network and 0.5 for small network.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>data.type</strong></td>
<td>Data representation used for module detection</td>
<td>- pearson: based on SCTransform (recommended) <br>- deviance: based on multinomial null model</td>
</tr>
<tr class="odd">
<td><strong>purity.optimization.step.size</strong></td>
<td>value of step size used for optimizing module purity. 0.5 recommended.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>filter.parameters</strong></td>
<td>specify which clusters to include or omit</td>
<td>named list with 2 sets of entries; ‘include’ entries specify which clusters to retain; ‘omit’ entries specify which clusters to omit. <br>E.g., list(include = c(0,1,2), omit = c(3,4,5))</td>
</tr>
<tr class="odd">
<td><strong>general.weight.by.var</strong></td>
<td>Weight feature embeddings by the variance of each PC. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>general.pca.cum.sum</strong></td>
<td>cumulative variance explained by PCs used for SSN analysis. Value between (0,1], 0.9 is recommended.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>general.umap.knn</strong></td>
<td>number of neighboring points used in local approximations of manifold structure. Larger values will result in more global structure being preserved at the loss of detailed local structure. In general this parameter should often be in the range 5 to 50.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>general.cluster.purity</strong></td>
<td>target purity of network modules. Value between [0,1], 0.8 is recommended.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>general.scale.free.topology</strong></td>
<td>apply scale-free transform to shared nearest neighbor graph. Default is TRUE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>general.ica</strong></td>
<td>perform independent component analysis (ICA).</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>general.nmf</strong></td>
<td>perform non-negative matrix factorization (NMF).</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>general.nmf.k</strong></td>
<td>number of latent factors to compute with NMF. Can be a range of values (e.g., c(5, 10, 15))</td>
<td>vector of numeric(s)</td>
</tr>
<tr class="odd">
<td><strong>general.robust.pca</strong></td>
<td>use robust PCA for dimensional reduction (<em>warning</em>: computationally intensive)</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>n.workers</strong></td>
<td>number of workers used for parallel implementations.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>print.inline</strong></td>
<td>print figures after each code chunk. Recommended if running script chunk-by-chunk. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>save.pdf</strong></td>
<td>save figures and tables</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>update.log</strong></td>
<td>update analysis log. Ignored if developer = F.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>rprofile.dir</strong></td>
<td>look for input file in path specified in .Rprofile. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>developer</strong></td>
<td>Developer mode. Default is FALSE.</td>
<td>logical</td>
</tr>
</tbody>
</table></details><p><br></p>
<hr>
</div>
</div>
<div id="module-06-gene-expression-and-associations" class="section level2">
<h2 class="hasAnchor">
<a href="#module-06-gene-expression-and-associations" class="anchor"></a>Module 06: Gene Expression and Associations</h2>
<div id="description-5" class="section level3">
<h3 class="hasAnchor">
<a href="#description-5" class="anchor"></a>Description</h3>
<p>Given a query gene or gene set, scRNAseq expression is evaluated and gene similarity metrics are computed to identify associated genes.</p>
</div>
<div id="example-reports-4" class="section level3">
<h3 class="hasAnchor">
<a href="#example-reports-4" class="anchor"></a>Example Reports</h3>
<ul>
<li><em>placeholder</em></li>
</ul>
</div>
<div id="inputs-and-outputs-4" class="section level3">
<h3 class="hasAnchor">
<a href="#inputs-and-outputs-4" class="anchor"></a>Inputs and Outputs</h3>
<details><p><summary><strong>show contents</strong></summary></p>
<p><u>Inputs</u></p>
<ul>
<li>
<strong>Seurat object</strong> (.Rds or .RData file)</li>
</ul>
<p><u>Outputs</u></p>
<ul>
<li>
<strong>Analysis report</strong> (.HTML file)</li>
</ul></details>
</div>
<div id="analysis-parameters-4" class="section level3">
<h3 class="hasAnchor">
<a href="#analysis-parameters-4" class="anchor"></a>Analysis Parameters</h3>
<details><p><summary><strong>show table</strong></summary></p>
<table class="table">
<colgroup>
<col width="11%">
<col width="55%">
<col width="33%">
</colgroup>
<thead><tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Argument</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>input.file</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>features</strong></td>
<td>features used for analysis. <br>If <code>feature.format</code> = 1, <code>features</code> must be vector of features present in seurat object (e.g., c(“PRRX1”, “TCF4”)). <br>If <code>feature.format</code> = 2, <code>features</code> must be name of column spreadsheet containing features (e.g., “HALLMARK_genes”). <br>If <code>feature.format</code> = 3, <code>features</code> argument is ignored.</td>
<td></td>
</tr>
<tr class="odd">
<td><strong>feature.format</strong></td>
<td>format of features provided by user.</td>
<td>- 0: user-provided genes <br>- 1: features are stored in spreadsheet. <br>- 2: discovery mode.</td>
</tr>
<tr class="even">
<td><strong>feature.csv.path</strong></td>
<td>Path to spreadsheet containing features. Required if using <code>feature.format</code> = 1. For intended use, features in spreadsheet must be organized in columns that can be referenced by the column header using the <code>features</code> argument.</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>discovery.method</strong></td>
<td>Differential expression algorithm used in discovery mode (<code>feature.format</code> = 2)</td>
<td>- Wilcoxon: Identifies DEGs using Wilcoxon Rank Sum test genes <br>- CDI: Identifies DEGs using co-dependency test <br>- Gini: Identifies group-specific features using gini inequality index</td>
</tr>
<tr class="even">
<td><strong>discovery.pct.min</strong></td>
<td>Features that are expressed below this value are omitted from differential expression analysis. Range between [0,1], 0.1 is recommended. Required only if <code>feature.format</code> = 2.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>discovery.group.by</strong></td>
<td>Name of object meta data field to group cells by when performing differential expression. E.g., “seurat_clusters”. Required only if <code>feature.format</code> = 2.</td>
<td>character</td>
</tr>
<tr class="even">
<td><strong>discovery.top.n</strong></td>
<td>path to seurat object</td>
<td>character</td>
</tr>
<tr class="odd">
<td><strong>discovery.max.n</strong></td>
<td>minimum fraction of genes required for gene set consideration. If fraction of genes present in seurat object is below this threshold, gene set is disqualified. Value between [0,1]. Default is 0.5.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>as.module</strong></td>
<td>aggregate features into single meta-program. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>cor.method</strong></td>
<td>feature association method</td>
<td>- pearson: Pearson correlation <br>- spearman: Spearman correlation <br>- rho_p: Proportionality constant <br>- cdi: co-dependency index</td>
</tr>
<tr class="even">
<td><strong>cor.top.n</strong></td>
<td>Number of top associated features to show. 70 is recommended.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>cor.which.data</strong></td>
<td>aslot to use for feature association analysis. “data” is recommended.</td>
<td>- data <br>- scale</td>
</tr>
<tr class="even">
<td><strong>cor.min.expr</strong></td>
<td>Features that are expressed below this value are omitted from feature association analysis. Range between [0,1], 0.025-0.1 is recommended.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>general.cluster.resolution</strong></td>
<td>Value of the resolution parameter, use a value above 1.0 if you want to obtain a larger number of communities.</td>
<td>numeric</td>
</tr>
<tr class="even">
<td><strong>general.subsample.factor</strong></td>
<td>Value between [0,1] specifying fraction of cells to subsample. Use 1 for no subsampling.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>general.subset</strong></td>
<td>Subset Seurat object according to specified meta data field. Only specified meta data entries are retained, while remaining of data is omitted. Set as “no.subset” for not subsetting.</td>
<td>data frame specifying which field (<code>df$field</code>) to subset on, and which field entries to retain (<code>df$subgroups</code>)</td>
</tr>
<tr class="even">
<td><strong>general.clean.clusters</strong></td>
<td>omit cells exceeding 3 median absolute deviations from cluster center in UMAP coordinate space.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>general.save.pdf</strong></td>
<td>save figures and tables</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>general.print.inline</strong></td>
<td>print figures after each code chunk. Recommended if running script chunk-by-chunk. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="odd">
<td><strong>general.rprofile.dir</strong></td>
<td>look for input file in path specified in .Rprofile. Default is FALSE.</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>general.barcode.recode</strong></td>
<td>Named list where names are new labels, and entries are old labels. If is provided for old labels, treated as wild card. e.g. list(in.vitro = c("invitro), in.vivo = NA). See <code>scMiko::recordBarcode(...)</code> for details.</td>
<td>Named list where names are new labels, and entries are old labels.</td>
</tr>
<tr class="odd">
<td><strong>expression.do.hex</strong></td>
<td>show UMAP expression using schex. Default is FALSE</td>
<td>logical</td>
</tr>
<tr class="even">
<td><strong>n.workers</strong></td>
<td>number of workers used for parallel implementation. Set to 1 for non-parallel run.</td>
<td>numeric</td>
</tr>
<tr class="odd">
<td><strong>developer</strong></td>
<td>Developer mode. Default is FALSE.</td>
<td>logical</td>
</tr>
</tbody>
</table></details><hr>
<details><p><summary><strong>Session Info</strong></summary></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-w64-mingw32/x64 (64-bit)
## Running under: Windows 10 x64 (build 19041)
## 
## Matrix products: default
## 
## locale:
## [1] LC_COLLATE=English_Canada.1252  LC_CTYPE=English_Canada.1252   
## [3] LC_MONETARY=English_Canada.1252 LC_NUMERIC=C                   
## [5] LC_TIME=English_Canada.1252    
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## loaded via a namespace (and not attached):
##  [1] rprojroot_2.0.2   digest_0.6.27     crayon_1.4.1      R6_2.5.1         
##  [5] jsonlite_1.7.2    formatR_1.11      magrittr_2.0.1    evaluate_0.14    
##  [9] stringi_1.7.4     rlang_0.4.11      cachem_1.0.6      fs_1.5.0         
## [13] jquerylib_0.1.4   bslib_0.3.0       ragg_1.1.3        rmarkdown_2.11   
## [17] pkgdown_1.6.1     textshaping_0.3.5 desc_1.3.0        tools_4.0.3      
## [21] stringr_1.4.0     yaml_2.2.1        xfun_0.26         fastmap_1.1.0    
## [25] compiler_4.0.3    systemfonts_1.0.2 memoise_2.0.0     htmltools_0.5.2  
## [29] knitr_1.36        sass_0.4.0</code></pre>
</details>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
