<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Load parameter specifications for Module 1 — analysisParameters • scMiko</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous" />


<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Load parameter specifications for Module 1 — analysisParameters" />
<meta property="og:description" content="Load parameter specification list for Module 1 (preprocessing and QC). If not specified, default settings are applied." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/install.html">Install</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Introductory Vignettes</li>
    <li>
      <a href="../articles/pbmc3k_tutorial.html">PBMC 3K guided tutorial</a>
    </li>
    <li>
      <a href="../articles/multimodal_vignette.html">Using Seurat with multi-modal data</a>
    </li>
    <li>
      <a href="../articles/spatial_vignette.html">Analysis, visualization, and integration of spatial datasets with Seurat</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Data Integration</li>
    <li>
      <a href="../articles/integration_introduction.html">Introduction to scRNA-seq integration</a>
    </li>
    <li>
      <a href="../articles/integration_mapping.html">Mapping and annotating query datasets</a>
    </li>
    <li>
      <a href="../articles/integration_rpca.html">Fast integration using reciprocal PCA (RPCA)</a>
    </li>
    <li>
      <a href="../articles/integration_large_datasets.html">Tips for integrating large datasets</a>
    </li>
    <li>
      <a href="../articles/atacseq_integration_vignette.html">Integrating scRNA-seq and scATAC-seq data</a>
    </li>
    <li>
      <a href="../articles/multimodal_reference_mapping.html">Multimodal reference mapping</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Load parameter specifications for Module 1</h1>
    
    <div class="hidden name"><code>analysisParameters.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Load parameter specification list for Module 1 (preprocessing and QC). If not specified, default settings are applied.</p>
    </div>

    <pre class="usage"><span class='fu'>analysisParameters</span><span class='op'>(</span>
  <span class='va'>which.data</span>,
  which.strata <span class='op'>=</span> <span class='cn'>NULL</span>,
  <span class='va'>organism.filter.flag</span>,
  <span class='va'>organism.include</span>,
  save.flag <span class='op'>=</span> <span class='cn'>T</span>,
  <span class='va'>save.filename</span>,
  save.directory <span class='op'>=</span> <span class='st'>"Preprocessed_Datasets/"</span>,
  set_names <span class='op'>=</span> <span class='cn'>NULL</span>,
  data.imputed.flag <span class='op'>=</span> <span class='cn'>F</span>,
  vars2regress <span class='op'>=</span> <span class='st'>"percent.mt"</span>,
  subsample_factor <span class='op'>=</span> <span class='fl'>1</span>,
  plt.log.flag <span class='op'>=</span> <span class='cn'>TRUE</span>,
  gene.upperlimit <span class='op'>=</span> <span class='fl'>9000</span>,
  gene.lowerlimit <span class='op'>=</span> <span class='fl'>200</span>,
  mt.upperlimit <span class='op'>=</span> <span class='fl'>60</span>,
  unmatched.rate.filter.flag <span class='op'>=</span> <span class='cn'>T</span>,
  do.additional.reduction <span class='op'>=</span> <span class='cn'>F</span>,
  cluster.resolution <span class='op'>=</span> <span class='fl'>0.4</span>,
  print.inline <span class='op'>=</span> <span class='cn'>FALSE</span>
<span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>which.data</th>
      <td><p>Character. Specify input data (input file specification are stored in separate data.frame. See Module 1 for details)</p></td>
    </tr>
    <tr>
      <th>which.strata</th>
      <td><p>Character. Specify how input data are subset. Leave unspecified to keep all data.</p></td>
    </tr>
    <tr>
      <th>organism.filter.flag</th>
      <td><p>Logical specifying whether inputs should be filtered by species. Recommended if multiple species are present in input.</p></td>
    </tr>
    <tr>
      <th>organism.include</th>
      <td><p>Character. Which species to include. One of:</p><ul>
<li><p>"Hs" - Human</p></li>
<li><p>"Mm" - Mouse</p></li>
<li><p>"Hs, Mm" - both species included</p></li>
</ul></td>
    </tr>
    <tr>
      <th>save.flag</th>
      <td><p>Logical specifying whether preprocessed data (seurat object) are saved as RData file. Default is TRUE</p></td>
    </tr>
    <tr>
      <th>save.filename</th>
      <td><p>Character. Output filename (must include .Rdata suffix)</p></td>
    </tr>
    <tr>
      <th>save.directory</th>
      <td><p>Character. output directory (Default is "Preprocessed Datasets/")</p></td>
    </tr>
    <tr>
      <th>set_names</th>
      <td><p>Optional character. Label for dataset.</p></td>
    </tr>
    <tr>
      <th>data.imputed.flag</th>
      <td><p>Logical specifying wehther data were imputed. Default is F.</p></td>
    </tr>
    <tr>
      <th>vars2regress</th>
      <td><p>Character vector specifying which parameters to regress during data scaling. Default is "percent.mt"</p></td>
    </tr>
    <tr>
      <th>subsample_factor</th>
      <td><p>Numeric [0,1]. Factor used to subsample data. Recommended during initial data exploration, especially if dataset is large.</p></td>
    </tr>
    <tr>
      <th>plt.log.flag</th>
      <td><p>Logical specifying whether QC violin plots are on log scale.</p></td>
    </tr>
    <tr>
      <th>gene.upperlimit</th>
      <td><p>Numeric. Upper limit for number of genes per cell. Default is 200.</p></td>
    </tr>
    <tr>
      <th>gene.lowerlimit</th>
      <td><p>Numeric. Lower limit for number of genes per cell. Default is 9000.</p></td>
    </tr>
    <tr>
      <th>mt.upperlimit</th>
      <td><p>Numeric [0,100]. Upper limit for mitochondrial percentage. Default is 60.</p></td>
    </tr>
    <tr>
      <th>unmatched.rate.filter.flag</th>
      <td><p>Logical flag for filtering cells unmatched rate. Adaptive threshold are used.</p></td>
    </tr>
    <tr>
      <th>do.additional.reduction</th>
      <td><p>Flag to generate ICA and TSNE reductions. Default is FALSE.</p></td>
    </tr>
    <tr>
      <th>cluster.resolution</th>
      <td><p>Resolution used for cell clustering.</p></td>
    </tr>
    <tr>
      <th>print.inline</th>
      <td><p>Logical specifying wheter data are printed in R notebook. Default is F - recommended if generated flexdashboard output.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>list</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://scholar.google.ca/scholar?hl=en&as_sdt=0%2C5&q=mikolajewicz+nicholas&oq='>Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


