<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Evaluate specificity of single-cell markers across several cluster resolutions. — multiSpecificity • scMiko</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Evaluate specificity of single-cell markers across several cluster resolutions. — multiSpecificity"><meta property="og:description" content="Evaluate specificity of single-cell markers across several cluster resolutions using co-dependency index-based specificity measure. Consider running multiCluster(...) first."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scMiko</a>
        <span class="version label label-danger" data-toggle="tooltip" data-placement="bottom" title="Unreleased version">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Getting Started
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">Computational Resources</li>
    <li>
      <a href="../articles/install.html">scMiko Package</a>
    </li>
    <li>
      <a href="../articles/scPipeline_getting_started.html">scPipeline Reports</a>
    </li>
  </ul></li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li class="dropdown-header">scMiko Vignettes</li>
    <li>
      <a href="../articles/DEG.html">Differential-expression using CDI</a>
    </li>
    <li>
      <a href="../articles/Cluster_Optimization.html">Selecting cluster resolution using specificity criterion</a>
    </li>
    <li>
      <a href="../articles/Marker_Catalog.html">Cell-type marker catalog</a>
    </li>
    <li>
      <a href="../articles/Cell_Annotation.html">Marker-based cell-type annotation using Miko Scoring</a>
    </li>
    <li>
      <a href="../articles/Module_Detection.html">Gene program discovery using SSN analysis</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">scPipeline Dashboard Reports</li>
    <li>
      <a href="../articles/scPipeline_module_overview.html">Overview of modules</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/NMikolajewicz/scMiko">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Evaluate specificity of single-cell markers across several cluster resolutions.</h1>
    
    <div class="hidden name"><code>multiSpecificity.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Evaluate specificity of single-cell markers across several cluster resolutions using co-dependency index-based specificity measure. Consider running multiCluster(...) first.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">multiSpecificity</span><span class="op">(</span>
  <span class="va">object</span>,
  <span class="va">cluster_names</span>,
  features <span class="op">=</span> <span class="cn">NULL</span>,
  deg_prefilter <span class="op">=</span> <span class="cn">T</span>,
  geosketch.subset <span class="op">=</span> <span class="cn">F</span>,
  cdi_bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,
  min.pct <span class="op">=</span> <span class="fl">0.1</span>,
  n.workers <span class="op">=</span> <span class="fl">1</span>,
  return_dotplot <span class="op">=</span> <span class="cn">T</span>,
  verbose <span class="op">=</span> <span class="cn">T</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>object</dt>
<dd><p>Seurat object with multi-resolution clusters provided in meta data.</p></dd>
<dt>cluster_names</dt>
<dd><p>Vector specifying names of all cluster configurations found in meta data.</p></dd>
<dt>features</dt>
<dd><p>If specified, marker specificity analysis is limited to specified features. Otherwise all features are used (more computationally intensive).</p></dd>
<dt>deg_prefilter</dt>
<dd><p>If TRUE, wilcoxon analysis is performed first to subset DEG features for downstream analysis. Results in faster performance. Default is TRUE.</p></dd>
<dt>geosketch.subset</dt>
<dd><p>Use GeoSketch method to subsample scRNA-seq data while preserving rare cell states (https://doi.org/10.1016/j.cels.2019.05.003). Logical, T or F (Default F). Recommended if cell type representation is imbalanced.</p></dd>
<dt>cdi_bins</dt>
<dd><p>Vector specifying binning for CDI-based specificity curve. Must range [0,1]. Default is seq(0, 1, by = 0.01).</p></dd>
<dt>min.pct</dt>
<dd><p>Minimal expression of features that are considered in specificity analysis. Represents fraction of expression cells and must range [0,1]. Higher values result in faster performance. Default is 0.1.</p></dd>
<dt>n.workers</dt>
<dd><p>Number of workers used for parallel implementation. Default is 1.</p></dd>
<dt>return_dotplot</dt>
<dd><p>If TRUE, dot plots visualizing expression of top specific markers are returned. Default is T.</p></dd>
<dt>verbose</dt>
<dd><p>Print progress. Default is TRUE.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Seurat object</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="multiCluster.html">multiCluster</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Nicholas Mikolajewicz</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># clustering data</span></span>
<span class="r-in"><span class="va">ms.list</span> <span class="op">&lt;-</span> <span class="fu">multiSpecificity</span><span class="op">(</span>object <span class="op">=</span> <span class="va">so.query</span>, cluster_names <span class="op">=</span> <span class="va">cluster.name</span>, features <span class="op">=</span> <span class="cn">NULL</span>, deg_prefilter <span class="op">=</span> <span class="cn">T</span>,</span>
<span class="r-in">cdi_bins <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, by <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>, min.pct <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span class="r-in">n.workers <span class="op">=</span> <span class="fl">4</span>, return_dotplot <span class="op">=</span> <span class="cn">T</span>,  verbose <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in "Seurat" %in% class(object):</span> object 'so.query' not found</span>
<span class="r-in"><span class="va">df.auc.spec</span> <span class="op">&lt;-</span> <span class="va">ms.list</span><span class="op">$</span><span class="va">specificity_summary</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'ms.list' not found</span>
<span class="r-in"><span class="va">qm.res.sum.sum.all</span> <span class="op">&lt;-</span> <span class="va">ms.list</span><span class="op">$</span><span class="va">specificity_raw</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'ms.list' not found</span>
<span class="r-in"><span class="va">plt.clust.spec</span> <span class="op">&lt;-</span> <span class="va">ms.list</span><span class="op">$</span><span class="va">auc_plot</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'ms.list' not found</span>
<span class="r-in"><span class="va">plt.auc.spec</span> <span class="op">&lt;-</span> <span class="va">ms.list</span><span class="op">$</span><span class="va">resolution_plot</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'ms.list' not found</span>
<span class="r-in"><span class="va">plt.auc.dot</span> <span class="op">&lt;-</span> <span class="va">ms.list</span><span class="op">$</span><span class="va">dot_plot</span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'ms.list' not found</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by <a href="https://scholar.google.ca/scholar?hl=en&amp;as_sdt=0%2C5&amp;q=mikolajewicz+nicholas&amp;oq=" class="external-link">Nicholas Mikolajewicz</a>.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

      </footer></div>

  


  

  </body></html>

